components/pedigree/resources/src/main/resources/pedigree/view/datepicker.js

```diff
@@ -1,276 +1,437 @@
+        this.container.insert(this.inputs[identifier].getElement());
+      }.bind(this));
+    },
+    updateDateFromInput : function() {
+      var dateObject = JSON.parse(this._input.value);
+      if (dateObject["year"] != null) {
+        this.inputs.y.setValue(dateObject["year"]);
+      } else if (dateObject["decade"] != null) {
+        this.inputs.y.setValue(dateObject["decade"]);
+      }
 
+      dateObject["month"] != null && this.inputs.m.setValue(dateObject["month"]);
+      dateObject["day"] != null && this.inputs.d.setValue(dateObject["day"]);
+      this.container.fire("datepicker:date:changed");
+    },
+    onDateChange : function(e) {
+      var validDate = true;
+      for (input in this.inputs) {
+        if (this.inputs.hasOwnProperty(input)){
+          this.inputs[input].onDateChange();
+          validDate = validDate && this.inputs[input].isValid;
+        }
+      }
+      if (validDate) {
+        this.updateDate(false);
+      }
+    },
     updateDate : function (doNotFireEventOnChange) {
         var dateObject = {};
 
-        var y = this.yearSelector.getSelectedValue();
+        var y = this.inputs.y.getValue();
         if (y.match(/\d\d\d\ds$/)) {
-            dateObject["decade"] = y;
+          dateObject["decade"] = y;
         } else {
-            if (y != "") {
-                dateObject["year"] = y;
-            }
+          if (y != "") {
+            dateObject["year"] = y;
+          }
         }
 
         if (y > 0 || this.inputFormat == "DMY") {
-            var m = this.monthSelector.getSelectedValue();
-            if (m > 0) {
-                dateObject["month"] = this.monthSelector.getSelectedOption();
-            }
+          var m = this.inputs.m.getValue();
+          if (m > 0) {
+            dateObject["month"] = m;
+          }
 
-            if (m > 0 || this.inputFormat == "DMY") {
-                var d = this.daySelector.getSelectedValue();
-                if (d > 0) {
-                    dateObject["day"] = this.daySelector.getSelectedOption();
-                }
+          if (m > 0 || this.inputFormat == "DMY") {
+            var d = this.inputs.d.getValue();
+            if (d > 0) {
+              dateObject["day"] = d;
             }
+          }
         }
 
         var newValue = JSON.stringify(dateObject);
-        if (newValue != this.__input.value) {
-            this.__input.value = JSON.stringify(dateObject);
-            if (!doNotFireEventOnChange) {
-                this.__input.fire("xwiki:date:changed");
-            }
+        if (newValue != this._input.value) {
+          this._input.value = JSON.stringify(dateObject);
+          if (!doNotFireEventOnChange) {
+            this._input.fire("xwiki:date:changed");
+          }
         }
+    },
+    initStrings: function() {
+      var strings = {
+        dayPlaceholder: 'DD',
+        monthPlaceholder: 'MM'
+      };
+      strings.yearPlaceholder = this.decadesEnabled ? 'YYYYs' : 'YYYY';
```

Comment:
should work as long as you initialize strings after you initialize this.decadesEnabled